app/scripts/migrations/120.3.ts

```diff
@@ -0,0 +1,110 @@
+ * result in some loss of information, but the impact is minor. The lost data
+ * is only used in the "Activity log" on the transaction details page.
+ *
+ * @param originalVersionedData - Versioned MetaMask extension state, exactly
+ * what we persist to dist.
+ * @param originalVersionedData.meta - State metadata.
+ * @param originalVersionedData.meta.version - The current state version.
+ * @param originalVersionedData.data - The persisted MetaMask state, keyed by
+ * controller.
+ * @returns Updated versioned MetaMask extension state.
+ */
+export async function migrate(
+  originalVersionedData: VersionedData,
+): Promise<VersionedData> {
+  const versionedData = cloneDeep(originalVersionedData);
+  versionedData.meta.version = version;
+  transformState(versionedData.data);
+  return versionedData;
+}
+
+function transformState(state: Record<string, unknown>) {
+  if (!hasProperty(state, 'TransactionController')) {
+    log.warn(`Migration ${version}: Missing TransactionController state`);
+    return state;
+  } else if (!isObject(state.TransactionController)) {
+    global.sentry?.captureException(
+      `Migration ${version}: Invalid TransactionController state of type '${typeof state.TransactionController}'`,
+    );
+    return state;
+  }
+
+  const transactionControllerState = state.TransactionController;
+
+  if (!hasProperty(transactionControllerState, 'transactions')) {
+    log.warn(
+      `Migration ${version}: Missing TransactionController transactions`,
+    );
+    return state;
+  } else if (!Array.isArray(transactionControllerState.transactions)) {
+    global.sentry?.captureException(
+      `Migration ${version}: Invalid TransactionController transactions state of type '${typeof transactionControllerState.transactions}'`,
+    );
+    return state;
+  }
+
+  const validTransactions =
+    transactionControllerState.transactions.filter(isObject);
+  if (
+    transactionControllerState.transactions.length !== validTransactions.length
+  ) {
+    const invalidTransaction = transactionControllerState.transactions.find(
+      (transaction) => !isObject(transaction),
+    );
+    global.sentry?.captureException(
+      `Migration ${version}: Invalid transaction of type '${typeof invalidTransaction}'`,
+    );
+    return state;
+  }
+
+  const validHistoryTransactions = validTransactions.filter(
+    hasValidTransactionHistory,
+  );
+  if (validHistoryTransactions.length !== validTransactions.length) {
+    const invalidTransaction = validTransactions.find(
+      (transaction) => !hasValidTransactionHistory(transaction),
+    );
+    global.sentry?.captureException(
+      `Migration ${version}: Invalid transaction history of type '${typeof invalidTransaction?.history}'`,
+    );
+    return state;
+  }
+
+  for (const transaction of validHistoryTransactions) {
+    if (transaction.history && transaction.history.length > 100) {
+      transaction.history = transaction.history.slice(0, 100);
```

Comment:
Wasn't the goal of the compression to retain later detail as we assume it's more relevant? Yet here we're prioritising the older history?

Reply:
Good question. In the case of compressing history, we were explicitly excluding any entries that we knew to be used for displaying the activity log. These displayed activities are predominantly at the beginning of the history, which is why we're preserving them here.

The preference for preserving newer non-displayed entries was for their use as diagnostic information. The thought was that for diagnosing strange behavior, recent changes might be more informative than earlier ones (the thought being that we'd be taking a snapshot of state shortly after something problematic happens). In this migration, we're risking discarding useful diagnostic information in favour of fixing the egregious performance problem.